lib/errors.js

```diff
@@ -0,0 +1,267 @@
+//
+// Required args:
+// - Exactly one of `args.exception` or `args.logMessage` must be set.
+//   `args.exception` is an Error instance. `args.logMessage` is a log message
+//   string, or an object of the form `{ message: 'template', params: [ ... ]}`
+//   which will be formated with `util.format()`.
+// - `args.log` {Logger}
+// - `args.shouldCaptureAttributes` {Boolean}
+// - `args.timestampUs` {Integer} - Timestamp of the error in microseconds.
+// - `args.handled` {Boolean}
+// - `args.sourceLinesAppFrames` {Integer} - Number of lines of source context
+//   to include in stack traces.
+// - `args.sourceLinesLibraryFrames` {Integer} - Number of lines of source
+//   context to include in stack traces. This and the previous arg are typically
+//   select from the agent `sourceLines{Error,Span}{App,Library}Frames` config
+//   vars.
+//
+// Optional args:
+// - `args.callSiteLoc` - A `Error.captureStackTrace`d object with a stack to
+//   include as `error.log.stacktrace`.
+// - `args.traceContext` - The current TraceContext, if any.
+// - `args.trans` - The current transaction, if any.
+// - `args.errorContext` - An object to be included as `error.context`.
+// - `args.message` - A message string that will be included as `error.log.message`
+//   if `args.exception` is given. Ignored if `args.logMessage` is given.
+//
+// This always calls back with `cb(null, apmError)`, i.e. it doesn't fail.
+function createAPMError (args, cb) {
+  let numAsyncStepsRemaining = 0 // finish() will call cb() only when this is 0.
+
+  const error = {
+    id: crypto.randomBytes(16).toString('hex'),
+    timestamp: args.timestampUs
+  }
+  if (args.traceContext) {
+    error.parent_id = args.traceContext.traceparent.id
+    error.trace_id = args.traceContext.traceparent.traceId
+  }
+  if (args.trans) {
+    error.transaction_id = args.trans.id
+    error.transaction = {
+      type: args.trans.type,
+      sampled: args.trans.sampled
+    }
+  }
+  if (args.errorContext) {
+    error.context = args.errorContext
+  }
+
+  if (args.exception) {
+    // Handle an exception, i.e. `captureError(<an Error instance>, ...)`.
+    const err = args.exception
+    const errMsg = String(err.message)
+    error.exception = {
+      message: errMsg,
+      type: String(err.name),
+      handled: args.handled
+    }
+
+    if ('code' in err) {
+      error.exception.code = String(err.code)
+    } else {
+      // To provide better grouping of mysql errors that happens after the async
+      // boundery, we modify to exception type to include the custom mysql error
+      // type (e.g. ER_PARSE_ERROR)
+      var match = errMsg.match(MYSQL_ERROR_MSG_RE)
+      if (match) {
+        error.exception.code = match[1]
+      }
+    }
+
+    // Optional add an alternative error message as well as the exception message.
+    if (args.message && typeof args.message === 'string') {
+      error.log = { message: args.message }
+    }
```

Comment:
REVIEW-NOTE: This is a slight change in guards on `opts.message` from before that I think more closely matches the API docs.  This changes the behaviour so that `error.log.message` is set whenever a message option is passed to `agent.captureError` like this:

```
agent.captureError(new Error('foo'), { message: 'foo' })
```

The implementation before was to have a special case where `error.log.message` was not added in that case *because err.message and opts.message were the same*. 

Here are the API docs on `captureError`:

> ** `message` - If the `error` argument is an +{type-error}+ object,
> it's possible to use this option to supply an additional message string that will be stored along with the error message under `log.message`

I think it is slightly surprising that a given `opts.message` might not get used if the err.message happens to be the same.

If you care strongly, I'm happy to revert to the old behaviour and tweak the API doc to be clearer.

Reply:
no strong opinions here, 👍 to the change.